README.md
# afnews

<!-- badges: start -->
[![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/armelsoubeiga/afnews?branch=master&svg=true)](https://ci.appveyor.com/project/armelsoubeiga/afnews)
[![lifecycle](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)
[![version](https://img.shields.io/github/tag/armelsoubeiga/afnews.svg)](https://github.com/armelsoubeiga/afnews/releases)
[![code-size](https://img.shields.io/github/languages/code-size/armelsoubeiga/afnews.svg)](https://github.com/armelsoubeiga/afnews)
[![R build status](https://github.com/armelsoubeiga/afnews/workflows/R-CMD-check/badge.svg)](https://github.com/armelsoubeiga/afnews/actions)
<!-- badges: end -->

<div align="center">

<img src="https://raw.githubusercontent.com/armelsoubeiga/afnews/master/logo.png" height="180px" />

Africa news data collection. A package to scraping all news of all countries of Africa. The news is by country and by date range. The data returned are: title, contents, comments, journal name and category. 

[Website](https://armelsoubeiga.github.io/afnews) | [Get Started](https://armelsoubeiga.github.io/afnews/get-started.html)

</div>

The goal of afnews is to ...

## Installation

You can install the released version of afnews from [GITHUB](https://github.com/armelsoubeiga/afnews/) with:

``` r
# install.packages("devtools")
devtools::install_github("armelsoubeiga/afnews")
```

## Example

This is a basic example which shows you how to solve a common problem:

``` r
library(afnews)
## basic example code
```



NEWS.md
# afnews 0.1.0

* Added a `NEWS.md` file to track changes to the package.


R/bf_news.R
#' Function of collecting articles on many press of Burkina Faso
#'
#' @description Collection function of all the articles of contry Burkina Faso
#' @param start_d yyyy-mm-dd start date of collection (oldest date, eg: yesterday)
#' @param end_d yyyy-mm-dd end date of collection (newest date, eg: today)
#' @param press specify the name of a press.
#' @return a table containing the articles published in the period
#'
#' @import xml2
#' @import rvest
#' @import tidyverse
#' @import lubridate
#' @import stringi
#'
#' @examples
#' library(afnews)
#' dt_ <- bf_news_collect(start_d="2020-10-20", end_d="2020-10-28", press="all")
#' head(dt_)
#' @export
bf_news_collect <- function(start_d, end_d, press="all"){

  ########
  ########  FASO NET ############
  fasonet <- function(start_d, end_d){
    #collect rubrique url
    url_rubrique <- read_html("https://lefaso.net/") %>%
                    html_nodes("a") %>%
                    html_attr("href") %>% str_trim() %>%
                    str_subset("^spip.php\\?rubrique")

  # Collect pagination by rubrique
    url_pag <- map_dfr(.x = unique(paste("http://lefaso.net/",url_rubrique,sep="")),
                       .f = function(x){
                        tibble(
                          url_pag=try(c(sub("http://lefaso.net/","",x),
                                       read_html(x) %>%
                                       html_nodes(".pagination a") %>%
                                       html_attr("href") %>% str_trim())))})
    url_pag$url_pag <- unique(paste("http://lefaso.net/",url_pag$url_pag,sep=""))

    # Collect article url by pagination url_category[-c(1,3)]
    art_fasonet <- data.frame()
    for(i in 1:length(url_pag$url_pag)){
      x <- url_pag$url_pag[i]

      date_article = try(
                      read_html(x) %>%
                      html_nodes("abbr") %>%
                      html_text())

      french.months <- c("janvier", "f\\u00e9vrier", "mars", "avril", "mai", "juin",
                         "juillet", "ao\\u00fbt", "septembre", "octobre", "novembre", "d\\u00e9cembre")
      date_article <- stringi::stri_replace_all_fixed(date_article, french.months, month.abb, vectorize_all=FALSE) %>% dmy() %>% date()



      if(is.na(tail(date_article, n=1L))){
        next()
      }else{
        d_i <- cbind_na(
                date_article = date_article,
                url_article = tryCatch(
                              read_html(x) %>%
                              html_nodes(".container .row div h3 a") %>%
                              html_attr("href") %>% str_trim(),
                              error = function(e){NA},
                              warning=  function(e){NA}),

               title_article = tryCatch(
                             read_html(x) %>%
                             html_nodes(".container .row div h3 a") %>%
                             html_text() %>% str_trim(),
                             error = function(e){NA},
                             warning=  function(e){NA}))

        art_fasonet <- rbind.data.frame(art_fasonet, d_i)

        if(tail(date_article, n=1L) > as.Date.character(start_d)-2){
          width <- options()$width
          cat(paste0(rep('>', width), collapse = ''), "\n")
          Sys.sleep(.05)
        }else{
          cat("Done!\n")
          break()
        }
      }
    }
    # url article
    art_fasonet$url_article <- paste0("http://lefaso.net/",art_fasonet$url_article)
    # return
    return(art_fasonet)
  }



  ########
  ########  BURKINA 24 ############
  burkina24 <- function(start_d, end_d){

    # category url collecte
    url_category <- read_html("https://www.burkina24.com/")%>%
                    html_nodes("#primary-menu .menu-inside > ul > li > a") %>%
                    html_attr("href") %>% str_trim() %>% .[-c(1,3)]

    # Collect pagination by rubrique
    number_pag <- map_dfr(.x = url_category,
                          .f = function(x){
                            tibble(url = x,
                                   nbr_pg=tryCatch(
                                           read_html(x) %>%
                                           html_node(".dots + .page-numbers") %>%
                                           html_text() %>% as.integer(),
                                           error = function(e){NA},
                                           warning=  function(e){NA})
                                           )})

    # Build all url
    number_pag <- na.omit(number_pag)
    number_pag_order <- number_pag %>%
                        split(1:nrow(.)) %>%
                        map_dfr(.f = function(x){
                          tibble(url = x$url,
                                 nbr_pg=1:x$nbr_pg)})
    number_pag_order <- number_pag_order[order(number_pag_order$nbr_pg),]
    url_pag_b24 <- paste0(number_pag_order$url,'page/', number_pag_order$nbr_pg)
    

    # Collect article url by pagination url_category[-c(1,3)]
    art_b24 <- data.frame()
    for(i in 1:length(url_pag_b24)){
          x <- url_pag_b24[i]
          date_article = try(
                        read_html(x) %>%
                        html_nodes(xpath = "//div[@class='posts-lists']/div/article") %>%
                        html_nodes(".meta > span:last-of-type") %>%
                        html_text())

          french.months <- c("janvier", "f\\u00e9vrier", "mars", "avril", "mai", "juin",
                             "juillet", "ao\\u00fbt", "septembre", "octobre", "novembre", "d\\u00e9cembre")
          date_article <- stringi::stri_replace_all_fixed(date_article, french.months, month.abb, vectorize_all=FALSE) %>% dmy() %>% date()


          if(is.na(tail(date_article, n=1L))){
            next()
          }else{
          d_i <- cbind_na(
            date_article = date_article,
            url_article = tryCatch(
                            read_html(x) %>%
                            html_nodes(xpath = "//div[@class='posts-lists']/div/article") %>%
                            html_nodes("h2 > a") %>% html_attr("href") %>% str_trim(),
                          error = function(e){NA},
                          warning=  function(e){NA}),
            title_article = tryCatch(
                            read_html(x) %>%
                            html_nodes(xpath = "//div[@class='posts-lists']/div/article") %>%
                            html_nodes("h2 > a") %>% html_text() %>% str_trim(),
                            error = function(e){NA},
                            warning=  function(e){NA}))

          art_b24 <- rbind.data.frame(art_b24, d_i)

          if(tail(date_article, n=1L) > as.Date.character(start_d)-2){
            width <- options()$width
            cat(paste0(rep('>', width), collapse = ''), "\n")
            Sys.sleep(.05)
          }else{
            cat("Done!\n")
            break()
          }
      }
    }
    # return
    return(art_b24)
}


  # Select article for time entries
  duration <- interval(as.Date.character(start_d), as.Date.character(end_d)) %>% as.numeric('days')

  if(duration > 30){
    base::warning("warning: the request interval is greater than 30 days.!")
    article_ <- NA
  }
  else{
    if(press=="fasonet"){
      article <- fasonet(start_d, end_d)
      article_ <- article %>% filter(between(date_article, as.Date.character(start_d), as.Date.character(end_d)))
    }else{
      if(press=="burkina24"){
        article <- burkina24(start_d, end_d)
        article_ <- article %>% filter(between(date_article, as.Date.character(start_d), as.Date.character(end_d)))
      }else{
        article_1 <- fasonet(start_d, end_d)
        article_2 <- burkina24(start_d, end_d)
        article_ <- rbind.data.frame(article_1,article_2) %>% filter(between(date_article, as.Date.character(start_d), as.Date.character(end_d)))
      }
    }
  }

  # verified if article show
  if(nrow(article_)==0){
    base::warning("No article publish in this period !")
  }

  return(article_)
}









#' Main functions of collecting news from Burkina Faso
#'
#' @description Function for collecting news (title + content) from the media in Burkina Faso.
#' @param start_d yyyy-mm-dd start date of collection (oldest date, eg: yesterday)
#' @param end_d yyyy-mm-dd collection end date (newest date, eg: today)
#' @param press specify the name of a press.
#' @return a table containing the news (title + content + date publisher + press) collected
#'
#' @import xml2
#' @import rvest
#' @import tidyverse
#' @import lubridate
#' @import rapportools
#'
#' @examples
#' library(afnews)
#' dt_news <- bf_news_contents(start_d="2020-10-20", end_d="2020-10-28", press='all')
#' head(dt_news)
#' @export
bf_news_contents <- function(start_d, end_d, press='all') {

  ######
  ###### FASONET CONTENT ###
  fasonet_news_contents <- function(start_d, end_d){

    fasonet_content <- bf_news_collect(start_d, end_d, press='fasonet')
    content_collect <- map_dfr(.x = as.character(fasonet_content$url_article),
                               .f = function(x){
                                    tibble(Content_article = tryCatch(
                                                            read_html(x) %>%
                                                            html_nodes(xpath = "//div[@class='article_content']") %>%
                                                            html_text()%>% str_trim(),
                                                            error = function(e){NA},
                                                            warning=  function(e){NA}),
                                                  press_article = "fasonet")})
    article_fasonet <- cbind.data.frame(fasonet_content,content_collect)
    return(article_fasonet)
  }

  ######
  ###### BURKINA 24 CONTENTS ###
  burkina24_news_contents <- function(start_d, end_d){

      burkina24_content <- bf_news_collect(start_d, end_d, press='burkina24')
      content_collect <- map_dfr(.x = as.character(burkina24_content$url_article),
                                 .f = function(x){
                                     tibble(Content_article = tryCatch(
                                                      read_html(x) %>%
                                                      html_nodes("article > .page-content") %>%
                                                      html_text()%>% str_trim(),
                                                      error = function(e){NA},
                                                      warning=  function(e){NA}),
                                            press_article = "burkina24")})
      article_burkina24 <- cbind.data.frame(burkina24_content,content_collect)
    return(article_burkina24)
  }



  #######
  ######### Call function
  if(is.empty(start_d) || is.empty(end_d) ){
    base::warning("`start_d` `end_d` and can not be empty !")
    article_dt <- NA
  }else{
  if(press=='fasonet'){
    article_dt <- fasonet_news_contents(start_d, end_d)
  }else{
    if(press=='burkina24'){
      article_dt <- burkina24_news_contents(start_d, end_d)
    }else{
      article_dt_fasonet <- fasonet_news_contents(start_d, end_d)
      article_dt_burkina24 <- burkina24_news_contents(start_d, end_d)
      article_dt <- rbind.data.frame(article_dt_fasonet,article_dt_burkina24)
    }
  }
}
  return(article_dt)
}



R/simdata.R
#' Simulate Multivariate Normal Distribution
#' @description Custom function to simulate multivariate normal distribution
#' @param n sample size (integer)
#' @param p number of variables (integer)
#' @param rho correlation between variables (between 0 and 1)
#' @return A nxp matrix of simulated data
#' @references
#' \url{https://gallery.rcpp.org/articles/simulate-multivariate-normal/}
#' @examples
#' dat <- mvrnormR(n = 100, p = 10, rho = 0.8)
#' heatmap(cor(dat))
#' @export
mvrnormR <- function(n, p, rho) {

  # covariance between Z_i and Z_j being rho^|i-j|
  times <- 1:p # used for creating covariance matrix
  H <- abs(outer(times, times, "-"))
  sigma <- rho^H

  mu <- rep(0, p)
  ncols <- ncol(sigma)
  mu <- rep(mu, each = n) ## not obliged to use a matrix (recycling)
  mu + matrix(stats::rnorm(n * ncols), ncol = ncols) %*% chol(sigma)
}


R/url_exists.R
#' Function to check if URL exists
#'
#' @param x a single URL
#' @param non_2xx_return_value what to do if the site exists but the
#'        HTTP status code is not in the `2xx` range. Default is to return `FALSE`.
#' @param quiet if not `FALSE`, then every time the `non_2xx_return_value` condition
#'        arises a warning message will be displayed. Default is `FALSE`.
#' @param ... other params (`timeout()` would be a good one) passed directly
#'        to `httr::HEAD()` and/or `httr::GET()`
#' @return return TRUE or FALSE
#'
#' @import httr
#'
#' @examples
#' urls <- c('http://www.amazon.com', 'http://this.isafakelink.biz', 'https://stackoverflow.com')
#' base::sapply(urls,url_exists)
#' @export
url_exists <- function(x, non_2xx_return_value = FALSE, quiet = FALSE,...){

  capture_error <- function(code, otherwise = NULL, quiet = TRUE) {
    tryCatch(
      list(result = code, error = NULL),
      error = function(e) {
        if (!quiet)
          message("Error: ", e$message)

        list(result = otherwise, error = e)
      },
      interrupt = function(e) {
        stop("Terminated by user", call. = FALSE)
      }
    )
  }

  safely <- function(.f, otherwise = NULL, quiet = TRUE) {
    function(...) capture_error(.f(...), otherwise, quiet)
  }

  sHEAD <- safely(httr::HEAD)
  sGET <- safely(httr::GET)

  # Try HEAD first since it's lightweight
  res <- sHEAD(x, ...)

  if (is.null(res$result) ||
      ((httr::status_code(res$result) %/% 200) != 1)) {

    res <- sGET(x, ...)

    if (is.null(res$result)) return(FALSE) # or whatever you want to return on "hard" errors

    if (((httr::status_code(res$result) %/% 200) != 1)) {
      if (!quiet) warning(sprintf("Requests for [%s] responded but without an HTTP status code in the 200-299 range", x))
      return(non_2xx_return_value)
    }

    return(TRUE)

  } else {
    return(TRUE)
  }

}




#' function for merging r objects (chars, int, data.frame, ...) of different sizes
#'
#' @description function for merging r objects (chars, int, data.frame, ...) of different sizes
#' @param ... R objects
#' @param first ysorting na values
#' @return a data frame
#'
#' @examples
#' library(afnews)
#' cbind_na(a=1:2, b=letters[1:5], c=c('a',1,3))
#' @export
cbind_na <- function(..., first = TRUE) {

  ## check lsit of entry
  padNA <- function (mydata, rowsneeded, first = TRUE){
    temp1 = colnames(mydata)
    rowsneeded = rowsneeded - nrow(mydata)
    temp2 = setNames(
      data.frame(matrix(rep(NA, length(temp1) * rowsneeded),
                        ncol = length(temp1))), temp1)
    if (isTRUE(first)) rbind(mydata, temp2)
    else rbind(temp2, mydata)
  }

  ## create cols names
  dotnames <- function(...) {
    vnames <- names(as.list(substitute(list(...)))[-1L])
    vnames <- unlist(vnames, FALSE, FALSE)
    vnames
  }

  ##main of function
  Names <- dotnames(...)
  datalist <- setNames(list(...), Names)
  nrows <- max(sapply(datalist, function(x)
    ifelse(is.null(dim(x)), length(x), nrow(x))))
  datalist <- lapply(seq_along(datalist), function(x) {
    z <- datalist[[x]]
    if (is.null(dim(z))) {
      z <- setNames(data.frame(z), Names[x])
    } else {
      if (is.null(colnames(z))) {
        colnames(z) <- paste(Names[x], sequence(ncol(z)), sep = "_")
      } else {
        colnames(z) <- paste(Names[x], colnames(z), sep = "_")
      }
    }
    padNA(z, rowsneeded = nrows, first = first)
  })
  do.call(cbind, datalist)
}



R/ml_news.R
#' Function of collecting articles on many press of Mali
#'
#' @description Collection function of all the articles of contry Mali
#' @param start_d yyyy-mm-dd start date of collection (oldest date, eg: yesterday)
#' @param end_d yyyy-mm-dd end date of collection (newest date, eg: today)
#' @param press specify the name of a press.
#' @return a table containing the articles published in the period
#'
#' @import xml2
#' @import rvest
#' @import tidyverse
#' @import lubridate
#' @import stringi
#'
#' @examples
#' library(afnews)
#' dt_ <- ml_news_collect(start_d="2020-10-20", end_d="2020-10-28", press="all")
#' head(dt_)
#' @export
ml_news_collect <- function(start_d, end_d, press="all"){

  ########
  ########  Lesechos ############
  echos <- function(start_d, end_d){
    #collect rubrique url
    url_rubrique <- read_html("https://www.lesechos.ml/") %>%
                    html_nodes("#header-layout2 .container #dropdown ul li a") %>%
                    html_attr("href") %>% str_trim() %>% .[-1]

    # Collect pagination by rubrique
    url_pag <- map_dfr(.x = url_rubrique,
                       .f = function(x){
                         tibble(
                           url_pag=try(read_html(x) %>%
                                         html_nodes(".pagination li a") %>%
                                         html_attr("href") %>% str_trim()))})
    url_pag <- unique(url_pag$url_pag)

    # Collect article url by pagination
    art_echos <- data.frame()
    for(i in 1:length(url_pag)){
      x <- url_pag[i]
      date_article = try(read_html(x) %>%
                           html_nodes(".bg-body .row .row .post-date-dark ul li:last-child") %>%
                           html_text()%>% str_trim())
      french.months <- c("janvier", "f\\u00e9vrier", "mars", "avril", "mai", "juin",
                         "juillet", "ao\\u00fbt", "septembre", "octobre", "novembre", "d\\u00e9cembre")
      date_article <- stringi::stri_replace_all_fixed(date_article, french.months, month.abb, vectorize_all=FALSE) %>% dmy() %>% date()

      if(is.na(tail(date_article, n=1L))){
        next()
      }else{
        d_i <- cbind_na(
            date_article = date_article,
            url_article = tryCatch(
                              read_html(x) %>%
                              html_nodes(".title-semibold-dark a") %>%
                              html_attr("href") %>% str_trim(),
                              error = function(e){NA},
                              warning=  function(e){NA}),
            title_article = tryCatch(
                              read_html(x) %>%
                              html_nodes(".title-semibold-dark a") %>%
                              html_text() %>% str_trim(),
                              error = function(e){NA},
                              warning=  function(e){NA}))
        art_echos <- rbind.data.frame(art_echos, d_i)

        if(tail(date_article, n=1L) > as.Date.character(start_d)-2){
          width <- options()$width
          cat(paste0(rep('>', width), collapse = ''), "\n")
          Sys.sleep(.05)
        }else{
          cat("Done!\n")
          break()
        }
      }
    }
    # return
    return(art_echos)
  }


  ########
  ########  maliweb ############
  maliweb <- function(start_d, end_d){

    # category url collecte
    url_category <- read_html("https://www.maliweb.net/")%>%
                    html_nodes("#td-header-menu .menu-td-demo-header-menu-container > ul > li > a") %>%
                    html_attr("href") %>% str_trim() %>% .[-c(1:2,9:11)]

    # Collect pagination by rubrique
    number_pag <- map_dfr(.x = url_category,
                          .f = function(x){
                            tibble(url = x,
                                   nbr_pg=tryCatch(
                                             read_html(x) %>%
                                             html_nodes(".page-nav .last") %>% html_text(),
                                             error = function(e){NA},
                                             warning=  function(e){NA})
                                        )})
    number_pag$nbr_pg <- as.numeric(gsub("[^0-9.-]+", "", as.character(number_pag$nbr_pg)))

    # Build all url
    number_pag <- na.omit(number_pag)
    number_pag_order <- number_pag %>%
                        split(1:nrow(.)) %>%
                        map_dfr(.f = function(x){
                          tibble(url = x$url,
                                 nbr_pg=1:x$nbr_pg)})
    number_pag_order <- number_pag_order[order(number_pag_order$nbr_pg),]
    url_pag_maliweb <- paste0(number_pag_order$url,'/page/', number_pag_order$nbr_pg)
    

    # Collect article url by pagination
    art_maliweb <- data.frame()
    for(i in 1:length(url_pag_maliweb)){
          x <- url_pag_maliweb[i]

          date_article = try(
                        read_html(x) %>%
                        html_nodes(".td-ss-main-content .item-details .td-module-meta-info .td-post-date time") %>%
                        html_text())

          french.months <- c("Jan", "F\\u00e9v", "Mar", "Avr", "Mai", "Juin",
                             "Juil", "Ao\\u00fbt", "Sep", "Oct", "Nov", "D\\u00e9c")
          date_article <- stringi::stri_replace_all_fixed(date_article, french.months, month.abb, vectorize_all=FALSE) %>% dmy() %>% date()


          if(is.na(tail(date_article, n=1L))){
            next()
          }else{
          d_i <- cbind_na(
                  date_article = date_article,
                  url_article = tryCatch(
                                  read_html(x) %>%
                                  html_nodes(".td-ss-main-content .item-details") %>%
                                  html_nodes("h3 > a") %>% html_attr("href") %>% str_trim(),
                                error = function(e){NA},
                                warning=  function(e){NA}),
                  title_article = tryCatch(
                                  read_html(x) %>%
                                  html_nodes(".td-ss-main-content .item-details") %>%
                                  html_nodes("h3 > a") %>% html_attr("title") %>% str_trim(),
                                  error = function(e){NA},
                                  warning=  function(e){NA}))

          art_maliweb <- rbind.data.frame(art_maliweb, d_i)

          if(tail(date_article, n=1L) > as.Date.character(start_d)-2){
            width <- options()$width
            cat(paste0(rep('>', width), collapse = ''), "\n")
            Sys.sleep(.05)
          }else{
            cat("Done!\n")
            break()
          }
      }
    }
    # return
    return(art_maliweb)
}


  # Select article for time entries
  duration <- interval(as.Date.character(start_d), as.Date.character(end_d)) %>% as.numeric('days')

  if(duration > 30){
    base::warning("warning: the request interval is greater than 30 days.!")
    article_ <- NA
  }
  else{
    if(press=="lesechos"){
      article <- echos(start_d, end_d)
      article_ <- article %>% filter(between(date_article, as.Date.character(start_d), as.Date.character(end_d)))
    }else{
      if(press=="maliweb"){
        article <- maliweb(start_d, end_d)
        article_ <- article %>% filter(between(date_article, as.Date.character(start_d), as.Date.character(end_d)))
      }else{
        article_1 <- echos(start_d, end_d)
        article_2 <- maliweb(start_d, end_d)
        article_ <- rbind.data.frame(article_1,article_2) %>% filter(between(date_article, as.Date.character(start_d), as.Date.character(end_d)))
      }
    }
  }

  # verified if article show
  if(nrow(article_)==0){
    base::warning("No article publish in this period !")
  }

  return(article_)
}









#' Main functions of collecting news from Mali
#'
#' @description Function for collecting news (title + content) from the media in Mali.
#' @param start_d yyyy-mm-dd start date of collection (oldest date, eg: yesterday)
#' @param end_d yyyy-mm-dd collection end date (newest date, eg: today)
#' @param press specify the name of a press.
#' @return a table containing the news (title + content + date publisher + press) collected
#'
#' @import xml2
#' @import rvest
#' @import tidyverse
#' @import lubridate
#' @import rapportools
#'
#' @examples
#' library(afnews)
#' dt_news <- ml_news_contents(start_d="2020-10-20", end_d="2020-10-28", press='all')
#' head(dt_news)
#' @export
ml_news_contents <- function(start_d, end_d, press='all') {

  ######
  ###### FASONET CONTENT ###
  echos_news_contents <- function(start_d, end_d){

    echos_content <- ml_news_collect(start_d, end_d, press='lesechos')

    title <-  map_dfr(.x = as.character(echos_content$url_article),
                      .f = function(x){
                          tibble(
                             title = tryCatch(
                                         read_html(x) %>%
                                         html_nodes(".bg-body .news-details-layout1 h2") %>%
                                         html_text()%>% str_trim(),
                                         error = function(e){NA},
                                         warning=  function(e){NA}))})
    echos_content$title_article = title$title
    content_collect <- map_dfr(.x = as.character(echos_content$url_article),
                               .f = function(x){
                                    tibble(Content_article = paste0(tryCatch(
                                                              read_html(x) %>%
                                                              html_nodes(".bg-body .news-details-layout1 p") %>%
                                                              html_text()%>% str_trim(),
                                                              error = function(e){NA},
                                                              warning=  function(e){NA}),
                                                              collapse = " "),
                                            press_article = "lesechos")})
    article_echos <- cbind.data.frame(echos_content,content_collect)
    return(article_echos)
  }


  ######
  ###### Mali web ###
  maliweb_news_contents <- function(start_d, end_d){

      maliweb_content <- ml_news_collect(start_d, end_d, press='maliweb')
      content_collect <- map_dfr(.x = as.character(maliweb_content$url_article),
                                 .f = function(x){
                                     tibble(Content_article = paste0(tryCatch(
                                                              read_html(x) %>%
                                                              html_nodes(".td-ss-main-content .td-post-content p") %>%
                                                              html_text()%>% str_trim(),
                                                              error = function(e){NA},
                                                              warning=  function(e){NA}),
                                                              collapse = " "),
                                            press_article = "maliweb")})
      article_maliweb<- cbind.data.frame(maliweb_content,content_collect)
    return(article_maliweb)
  }



  #######
  ######### Call function
  if(is.empty(start_d) || is.empty(end_d) ){
    base::warning("`start_d` `end_d` and can not be empty !")
    article_dt <- NA
  }else{
  if(press=='lesechos'){
    article_dt <- echos_news_contents(start_d, end_d)
  }else{
    if(press=='maliweb'){
      article_dt <- maliweb_news_contents(start_d, end_d)
    }else{
      article_dt_echos <- echos_news_contents(start_d, end_d)
      article_dt_maliweb <- maliweb_news_contents(start_d, end_d)
      article_dt <- rbind.data.frame(article_dt_echos,article_dt_maliweb)
    }
  }
}
  return(article_dt)
}

